test/utils.py

```diff
@@ -0,0 +1,819 @@
+                        "--protocol tcp "
+                        f"--destination {self.host} "
+                        f"--destination-port {self.port} "
+                        "--jump DROP "
+                    )
+                elif BSD:
+                    sudo(f"pfctl -a pgbouncer_test/port_{self.port} -F all")
+
+    @contextmanager
+    def reject_traffic(self):
+        """Rejects all traffic to this query runner with a TCP RST message"""
+        with self.enable_firewall():
+            if LINUX:
+                sudo(
+                    "iptables --append OUTPUT "
+                    "--protocol tcp "
+                    f"--destination {self.host} "
+                    f"--destination-port {self.port} "
+                    "--jump REJECT "
+                    "--reject-with tcp-reset"
+                )
+            elif BSD:
+                sudo(
+                    "bash -c '"
+                    f'echo "block return-rst out out proto tcp from any to {self.host} port {self.port}"'
+                    f"| pfctl -a pgbouncer_test/port_{self.port} -f -'"
+                )
+            else:
+                raise Exception("This OS cannot run this test")
+            try:
+                yield
+            finally:
+                if LINUX:
+                    sudo(
+                        "iptables --delete OUTPUT "
+                        "--protocol tcp "
+                        f"--destination {self.host} "
+                        f"--destination-port {self.port} "
+                        "--jump REJECT "
+                        "--reject-with tcp-reset"
+                    )
+                elif BSD:
+                    sudo(f"pfctl -a pgbouncer_test/port_{self.port} -F all")
+
+
+class Postgres(QueryRunner):
+    def __init__(self, pgdata):
+        self.port_lock = PortLock()
+        super().__init__("127.0.0.1", self.port_lock.port)
+        self.pgdata = pgdata
+        self.log_path = self.pgdata / "pg.log"
+        self.connections = {}
+        self.cursors = {}
+
+    def initdb(self):
+        run(
+            f"initdb -A trust --nosync --username postgres --pgdata {self.pgdata}",
+            stdout=subprocess.DEVNULL,
+        )
+
+        with self.conf_path.open(mode="a") as pgconf:
+            if USE_UNIX_SOCKETS:
+                pgconf.write("unix_socket_directories = '/tmp'\n")
+            pgconf.write("log_connections = on\n")
+            pgconf.write("log_disconnections = on\n")
+            pgconf.write("logging_collector = off\n")
+            # We need to make the log go to stderr so that the tests can
+            # check what is being logged.  This should be the default, but
+            # some packagings change the default configuration.
+            pgconf.write("log_destination = stderr\n")
+            # This makes tests run faster and we don't care about crash safety
+            # of our test data.
+            pgconf.write("fsync = false\n")
```

Comment:
This is new btw, but it changed test runtime from ~33s to ~22s on my machine. And it seemed like a small enough addition to bundle it in here.